File: HOT100/13.2/215.数组中的第k个最大元素.py
#
# @lc app=leetcode.cn id=215 lang=python3
#
# [215] 数组中的第K个最大元素
#
from mytools import *
# @lc code=start
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        def dfs(pack, k):
            l = []; m = []; r = []
            piv = random.choice(pack)
            for x in pack:
                if x < piv:
                    l.append(x)
                elif x == piv:
                    m.append(x)
                else:
                    r.append(x)
            if len(r) >= k:
                return dfs(r, k)
            elif len(r) + len(m) < k:
                return dfs(l, k - (len(r) + len(m)))
            else: return piv
        return dfs(nums, k)
# @lc code=end

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        h = []
        n = len(nums)
        for i in range(n):
            if len(h) < k: heapq.heappush(h, nums[i])
            else: heapq.heappushpop(h, nums[i])
        return h[0]

File: HOT100/HOT100p1/13.堆/215.数组中的第k个最大元素.cpp
/*
 * @lc app=leetcode.cn id=215 lang=cpp
 *
 * [215] 数组中的第K个最大元素
 */
#include "tools.h"
// @lc code=start
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        
        auto dfs = [&](auto&& self, vector<int>& nums, int k) -> int
        {
            vector<int> l, m, r;
            int piv = nums[rand() % nums.size()];
            for (auto x : nums)
            {
                if (x > piv) r.push_back(x);
                else if (x == piv) m.push_back(x);
                else l.push_back(x);
            }
            if (k <= r.size()) return self(self, r, k);
            if (k > r.size() + m.size()) return self(self, l, k - (r.size() + m.size()));
            return piv;
        };

        return dfs(dfs, nums, k);
    }
};
// @lc code=end

